[expo-notifications SDK39] Crash when Proguard is enabled

## 🐛 Bug Report

### Summary of Issue <!-- (just a few sentences) -->
Crash when Proguard is enabled

I think that this bug is related to this PR https://github.com/expo/expo/pull/9731 of @lukmccall 

Logcat:
```
2020-09-27 12:33:50.467 31954-31976/? E/unknown:DisabledDevSupportManager: Caught exception
    java.lang.IllegalStateException: Unable to instantiate AppLoader!
        at expo.modules.notifications.d.l.i.a(Unknown Source:98)
        at expo.modules.notifications.d.k.b.<init>(Unknown Source:17)
        at expo.modules.notifications.a.c(Unknown Source:30)
        at org.unimodules.adapters.react.e.b(Unknown Source:42)
        at org.unimodules.adapters.react.b.a(Unknown Source:2)
        at com.facebook.react.v.a(Unknown Source:41)
        at com.facebook.react.e.a(Unknown Source:30)
        at com.facebook.react.q.a(Unknown Source:34)
        at com.facebook.react.q.a(Unknown Source:54)
        at com.facebook.react.q.a(Unknown Source:36)
        at com.facebook.react.q.a(Unknown Source:0)
        at com.facebook.react.q$e.run(Unknown Source:68)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.ClassNotFoundException: expo.modules.notifications.NotificationsScoper
        at java.lang.Class.classForName(Native Method)
        at java.lang.Class.forName(Class.java:453)
        at java.lang.Class.forName(Class.java:378)
        at expo.modules.notifications.d.l.i.a(Unknown Source:39)
        at expo.modules.notifications.d.k.b.<init>(Unknown Source:17) 
        at expo.modules.notifications.a.c(Unknown Source:30) 
        at org.unimodules.adapters.react.e.b(Unknown Source:42) 
        at org.unimodules.adapters.react.b.a(Unknown Source:2) 
        at com.facebook.react.v.a(Unknown Source:41) 
        at com.facebook.react.e.a(Unknown Source:30) 
        at com.facebook.react.q.a(Unknown Source:34) 
        at com.facebook.react.q.a(Unknown Source:54) 
        at com.facebook.react.q.a(Unknown Source:36) 
        at com.facebook.react.q.a(Unknown Source:0) 
        at com.facebook.react.q$e.run(Unknown Source:68) 
        at java.lang.Thread.run(Thread.java:764) 
     Caused by: java.lang.ClassNotFoundException: Didn't find class "expo.modules.notifications.NotificationsScoper" on path: DexPathList[[zip file "/data/app/com.expobare--PgJMNQvILLRpH639aMniA==/base.apk"],nativeLibraryDirectories=[/data/app/com.expobare--PgJMNQvILLRpH639aMniA==/lib/arm64, /data/app/com.expobare--PgJMNQvILLRpH639aMniA==/base.apk!/lib/arm64-v8a, /system/lib64, /vendor/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:125)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at java.lang.Class.classForName(Native Method) 
        at java.lang.Class.forName(Class.java:453) 
        at java.lang.Class.forName(Class.java:378) 
        at expo.modules.notifications.d.l.i.a(Unknown Source:39) 
        at expo.modules.notifications.d.k.b.<init>(Unknown Source:17) 
        at expo.modules.notifications.a.c(Unknown Source:30) 
        at org.unimodules.adapters.react.e.b(Unknown Source:42) 
        at org.unimodules.adapters.react.b.a(Unknown Source:2) 
        at com.facebook.react.v.a(Unknown Source:41) 
        at com.facebook.react.e.a(Unknown Source:30) 
        at com.facebook.react.q.a(Unknown Source:34) 
        at com.facebook.react.q.a(Unknown Source:54) 
        at com.facebook.react.q.a(Unknown Source:36) 
        at com.facebook.react.q.a(Unknown Source:0) 
        at com.facebook.react.q$e.run(Unknown Source:68) 
        at java.lang.Thread.run(Thread.java:764) 
```

### Environment - output of `expo diagnostics` & the platform(s) you're targeting

```
  Expo CLI 3.27.10 environment info:
    System:
      OS: macOS 10.15.5
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 14.4.0 - /usr/local/bin/node
      Yarn: 1.19.1 - /usr/local/bin/yarn
      npm: 6.14.5 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.9.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.0, DriverKit 19.0, macOS 10.15, tvOS 14.0, watchOS 7.0
      Android SDK:
        Android NDK: 21.3.6528147
    IDEs:
      Android Studio: 4.0 AI-193.6911.18.40.6514223
      Xcode: 12.0/12A7209 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~39.0.2 => 39.0.3
      react: 16.13.1 => 16.13.1
      react-dom: 16.13.1 => 16.13.1
      react-native: ~0.63.2 => 0.63.2
      react-native-web: ~0.13.12 => 0.13.14
      expo-notifications: ~0.7.2
    npmGlobalPackages:
      expo-cli: 3.27.10
    Expo Workflow: bare
```
 
### Steps to Reproduce

1. `expo init ExpoBare`
2. select Bare workflow -> minimal
3. `cd ExpoBare`
4. `expo install expo-notifications`
5. Open project in Android Studio
6. Enable Proguard (update `enableProguardInReleaseBuilds` from `false` to `true` in `app/build.gradle` file)
7. Create release build
8. Run apk on device
9. The app crash on start


### Expected Behavior vs Actual Behavior

Expected Behavior: release build works great
Actual Behavior: release build crash

--- a/android/app/proguard-rules.pro
+++ b/android/app/proguard-rules.pro
@@ -24,7 +24,6 @@
 -dontnote **
 
 -keep class host.exp.exponent.generated.AppConstants { *; }
--keep class host.exp.exponent.taskManager.ExpoHeadlessAppLoader { *; }
 
 ##### Crashlytics #####
 -keepattributes SourceFile,LineNumberTable

--- a/android/expoview/src/main/java/host/exp/exponent/taskManager/ExpoHeadlessAppLoader.java
+++ b/android/expoview/src/main/java/host/exp/exponent/taskManager/ExpoHeadlessAppLoader.java
@@ -14,6 +14,7 @@ import java.util.Map;
 
 import host.exp.exponent.headless.InternalHeadlessAppLoader;
 
+@DoNotStrip
 public class ExpoHeadlessAppLoader implements HeadlessAppLoader {
 
   private static final String TAG = "TaskManagerInternalAppL";

--- a/packages/@unimodules/core/CHANGELOG.md
+++ b/packages/@unimodules/core/CHANGELOG.md
@@ -10,6 +10,8 @@
 
 ### 🐛 Bug fixes
 
+- Fixed the `DoNotStrip` annotation not working with classes. ([#10421](https://github.com/expo/expo/pull/10421) by [@lukmccall](https://github.com/lukmccall))
+
 ## 5.5.0 — 2020-08-11
 
 ### 🎉 New features

--- a/packages/@unimodules/core/android/proguard-rules.pro
+++ b/packages/@unimodules/core/android/proguard-rules.pro
@@ -7,6 +7,7 @@
   @org.unimodules.core.interfaces.ExpoMethod *;
 }
 
+-keep @org.unimodules.core.interfaces.DoNotStrip class *
 -keepclassmembers class * {
   @org.unimodules.core.interfaces.DoNotStrip *;
 }

--- a/packages/expo-notifications/CHANGELOG.md
+++ b/packages/expo-notifications/CHANGELOG.md
@@ -13,6 +13,7 @@
 - Fixed case where Android apps could crash if you set a new category with a text input action **without** providing any `options`. ([#10141](https://github.com/expo/expo/pull/10141) by [@cruzach](https://github.com/cruzach))
 - Android apps no longer rely on the `submitButtonTitle` property as the action button title (they rely on `buttonTitle`, which matches iOS behavior). ([#10141](https://github.com/expo/expo/pull/10141) by [@cruzach](https://github.com/cruzach))
 - Fixed `Notifications.requestPermissions()` returning `undetermined` instead of a known status in some browsers. ([#10296](https://github.com/expo/expo/pull/10296) by [@sjchmiela](https://github.com/sjchmiela))
+- Fixed crashing when Proguard is enabled. ([#10421](https://github.com/expo/expo/pull/10421) by [@lukmccall](https://github.com/lukmccall))
 
 ## 0.7.1 — 2020-08-26
 

--- a/packages/expo-notifications/android/src/main/java/expo/modules/notifications/NotificationsScoper.java
+++ b/packages/expo-notifications/android/src/main/java/expo/modules/notifications/NotificationsScoper.java
@@ -1,10 +1,13 @@
 package expo.modules.notifications;
 
+import org.unimodules.core.interfaces.DoNotStrip;
+
 import expo.modules.notifications.notifications.interfaces.NotificationsBuilderCreator;
 import expo.modules.notifications.notifications.interfaces.NotificationsReconstructor;
 import expo.modules.notifications.notifications.model.BareExpoNotificationsReconstructor;
 import expo.modules.notifications.notifications.presentation.builders.CategoryAwareNotificationBuilder;
 
+@DoNotStrip
 public class NotificationsScoper implements expo.modules.notifications.notifications.interfaces.NotificationsScoper {
 
   @Override

